language-plutus-core/plutus-ir/Language/PlutusIR/Transform/Inline.hs

```diff
@@ -0,0 +1,109 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs            #-}
+{-# LANGUAGE LambdaCase       #-}
+{-# LANGUAGE TypeFamilies     #-}
+{-# LANGUAGE TypeOperators    #-}
+{-|
+A simple inlining pass.
+
+The point of this pass is mainly to tidy up the code, not to particularly optimize performance.
+In particular, we want to get rid of "trivial" let bindings which the Plutus Tx compiler sometimes creates.
+-}
+module Language.PlutusIR.Transform.Inline where
+
+import           Language.PlutusIR
+import           Language.PlutusIR.Transform.Substitute
+
+import           Language.PlutusCore.Name
+
+import           Control.Lens                           hiding (Strict)
+
+{- Note [Repeated inlining]
+Consider the following program:
+
+let x = 1 in let y = x in y
+
+If we naively inlined both bindings, we might end up with
+
+let x = 1 in let y = 1 in x
+
+What happened? The substitution y->x itself mentions a variable we want to
+inline!
+
+There are various ways to fix this, but ultimately they all involve doing
+substitutions multiple times (and hence multiple tree traversals). We'd ideally
+like to minimize this.
+
+The approach we take is to:
+- Collect all the substitutions we want to do in the whole program (needs
+global uniqueness).
+- Apply all the substitutions bottom up, at each point applying the substitution
+to the substituted term until we can't do any more.
```

Comment:
I think your particular case would be neater as a graph algorithm, perhaps? But that would only work when you only care about variables.

I think you could also do something like: whenever you add a substitution, also run it over the RHSs of all the existing substitutions (assuming you're going inside-out). I think that ends up being close to what you suggest. It does involve a bunch of extra term traversal, but only the RHSs of lets, rather than the whole tree. So maybe it's better.

Reply:
Revised suggestion: also store the free variables of the substitution RHS along with the substitution. Then you only need to reprocess the RHS if the newly added substitution is for a variable that's free in it. That should get rid of almost all the re-traversal.

Reply:
Surely it must be well-known how to do this?

Reply:
That's a very good point. I don't remember seeing any discussion of this sort of problem but I'll have another look.